### DISCUSSION TITLE
Git branches - what goes where?

### DISCUSSION THREAD

**Discussion by introt** [zim-desktop-wiki/zim-desktop-wiki#1860] [Development]
Currently, the default branch is the `master` branch. While the `CONTRIBUTING.md` refers to a development branch, this branch is not explicitly named, and it is not mentioned under neither the "Bug fixes" or "Pull requests" headings.

Combined, these two lead to pull requests being opened against the master branch instead of the `develop` branch, leading to duplicate efforts (#1856) and the occasional conflict due to `develop` getting out-of-sync with `master` (which can also happen due to hot fixes).

* Should `master` always be reserved for the latest stable release and PRs be opened towards `develop`?
* Should `develop` be made the default branch?
  * would fix Issues' linked PRs, but also close them on merge, making open issues reflect the development branch instead of the latest release

The answers to these questions should be reflected in both the documentation and pull request templates -- just tell me which branch to open a PR against ;)

**Comment by introt** [comment]
Regarding `master` vs. `main`, we could also go for eg. `stable` or `release` etc. to better convey its intended purpose

**Comment by hugovk** [comment]
https://github.com/zim-desktop-wiki/zim-desktop-wiki/pull/1856 author here :wave:

Whichever branch you want people to make PRs against, please set that as the default branch in GitHub settings.

That way, GitHub presents that when visiting the site, and for cloning, and most importantly as the target branch for PRs.

So in my case I would have seen it would already have been fixed, and if not, I would have made the PR to the correct branch. Thanks!

**Reply by jaap-karssenberg** [reply]
Would it make sense to set the "develop" branch as the default branch and rename the "master" branch to "release" ?

I think that best reflects the current way is working.

**Reply by introt** [reply]
@jaap-karssenberg Yes, sounds good to me :+1:

**Reply by dehesselle** [reply]
@jaap-karssenberg Can you perform the necessary changes then? At least setting the default branch to `develop` would be very helpful as [that is required](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#about-linked-issues-and-pull-requests) for a PR to close a linked issue.

**Comment by jaap-karssenberg** [comment]
Yes I have this on my list. However, the wow with issues is not what I
would want, actually prefer the current way. Now a ticket is closed when I
merge into "master", which I do at release, so user that search issues find
open stuff compared to last release. When closing tickets as soon as they
go in "develop" means users do not get to see those when they search for
open issues even though the fix is not yet released.

Not sure how other projects do this.

**Reply by dehesselle** [reply]
Let me tell you about a common and widely adopted pattern that also plays into the strengths of issue management.

- Get rid of the `develop` branch. The `master` branch will act as a development branch.
- Develop features/fixes/... in dedicated (and usually short-lived) feature branches. (Just like outside contributors are already doing when they submit PRs.)
- Once a feature/fix/... is ready, merge it to `master`.
- Make use of "close on commit" and similar functionalities. Issues are supposed to be closed once the code hits `master`.
- Contributing and issue reporting guidelines need to be amended with something like this:
  > Before you report an issue, check if you can still reproduce it using latest code from master branch - it might have been already fixed. If not, check the issue tracker if someone else has already reported it. Otherwise, please open a new issue.
  - Since Zim can be run from source, there is no disproportionate effort required to test the latest code.
  - On the other hand (or for binary stuff), this is where nightly builds or CI artifacts come into play.
- When you reach the point of being feature complete in `master`  for a new release, create a release branch (e.g. `zim_0.75`). This will be a permanent branch.
  - "Feature complete" does not imply "all bugs fixed"! On the contrary, it actually only acts as feature freeze. From here on, only bugs that need fixing before the actual release will be developed and committed to the release branch.
  - Cherry-pick fixes from release branch to `master` where necessary.
  - Meanwhile, development of new features/breaking changes/... continues in `master` as explained above, without interruption.
  - Once you're satisfied with a soon-to-be release, create the release tag in the release branch.

**Comment by shawnhcorey** [comment]
On 2022-08-11 13:52, René de Hesselle wrote:
> # Develop features/fixes/... in dedicated (and usually short-lived) 
> feature branches. (Just like outside contributors are already doing 
> when they submit PRs.)
> # Once a feature/fix/... is ready, merge it to |master|.

One thing the Linux Foundation does is write test plans that demonstrate 
the bug before any code is changed. The tests are added to its test 
suite so that the bug will never appear again.

**Reply by dehesselle** [reply]
That is a good pattern as well, but it would dramatically increase the burden (and would not find any acceptance here, IMHO) for what is essentially a one-man-project with a few supporters around it.

**Comment by jaap-karssenberg** [comment]
I've gone ahead and switched the "develop" branch to be the default branch.
Also I renamed "master" to "release".

Key is that the "develop" branch is tracking the latest & greatest while
"release" is a placeholder for the last released version (and some
packaging fixes).

Agree that in theory the "release" branch is not need, since releases are
tagged in git, but keeping it for now. May delete it later.

The "develop" branch is the source for all new releases. I agree feature
branches are a good way of working and I should adopt this pattern myself
as well. Basically all PRs are feature branches that are then merged into
"develop". It is just that I do most of my own patches directly on this
branch without a PR. But I'm conscious about not pushing half finished
work. So in theory the "develop" branch should always be in a working state.

With respect to test cases, I often do add test cases for bug fixes, but
usually only for complex bugs not the trivial ones. I think this is
acceptable as it are the complex ones that have a risk of returning.
Trivial bugs will happen but do not often return in the same form.

--- Jaap

**Reply by dehesselle** [reply]
I don't mind if you (as the only one) keep pushing directly to `develop` if you keep it in a working state (as you just said). I find it more important that we can rely more on how GitHub intends us to work and don't have to artificially keep issues open or similar, which is now possible with `develop` being the default branch.
